### PR TITLE
fix($resource): fulfill promise with the correct value on error

### LINKED ISSUE #14837: bug($resource): promise not resolved with value returned by `responseErrorInterceptor`
**Do you want to request a *feature* or report a *bug*?**
Bug.

**What is the current behavior?**
Since 71cf28c (landed in master only), when there is either a `responseError` interceptor, the promise returned by a `$resource` instance action (or exposed as `instance.$promise`) is always resolved with the original response, instead of being fulfilled (resolved or rejected) with the value returned by the interceptor.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**
[Demo][1] (running with `1.5.7` - change versions to `snapshot` to see the issue)
- `1.5.7`: The promise is rejected with the value returned by the `responseError` interceptor.
- `snapshot`: The promise is resolved (instead of rejected) with the original response (instead of the value returned by the `responseError` interceptor).

**What is the expected behavior?**
Behavior on `snapshot` (eventual 1.6.x) should be the same as on `1.5.x`.

**What is the motivation / use case for changing the behavior?**
Fix a regression.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
Works on `1.5.x`. Broken on `snapshot` (eventual 1.6.x).

[1]: https://codepen.io/ExpertSystem/pen/mEWpEL?editors=0012

### PR DESCRIPTION
**Pull request by gkalpak**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.

**What is the current behavior? (You can also link to an open issue here)**
See #14837.

**What is the new behavior (if this is a feature change)?**
Restores the 1.5.x behavior.

**Does this PR introduce a breaking change?**
No, since it fixes (yet unreleased) regression.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

**Other information**:
Fixes #14837

The first commit (which only contains a couple of tests) could be backported to 1.5.x, in order to verify that the behavior is correct on that branch.

### CODE REVIEW

File: src/ngResource/resource.js [hunk 1]
```diff
@@ -776,13 +778,18 @@ angular.module('ngResource', ['ng']).
                 (success || noop)(value, response.headers);
                 return value;
               },
-              responseErrorInterceptor || error ?
+              (hasError || hasResponseErrorInterceptor) ?
```
Comment (Narretz) [inline src/ngResource/resource.js]: Can we use a real `if` here? The whole block is pretty hard to read (for me)

File: src/ngResource/resource.js [hunk 2]
```diff
@@ -776,13 +778,18 @@ angular.module('ngResource', ['ng']).
...
-              }
-              : undefined);
+                if (hasError) error(response);
+                return hasResponseErrorInterceptor ?
+                    responseErrorInterceptor(response) :
+                    $q.reject(response);
+              } :
+              undefined);
+            if (hasError && !hasResponseErrorInterceptor) {
+              // Avoid PUR, but still reject the returned promise
```
Comment (Narretz) [inline src/ngResource/resource.js]: PUR?

Comment (gkalpak) [inline src/ngResource/resource.js]: You know. **P**ossibly **U**nhandled **R**ejection :stuck_out_tongue:

Comment (petebacondarwin) [inline src/ngResource/resource.js]: false economy :-P

Comment (gkalpak) [inline src/ngResource/resource.js]: Fixed.

**Pull request closed without merging**